src/webgpu/shader/validation/expression/access/vector.spec.ts

```diff
@@ -0,0 +1,203 @@
+  xyxy: { wgsl: 'let r : vec4<T> = v.xyxy;', ok: true },
+  xyxz: { wgsl: 'let r : vec4<T> = v.xyxz;', ok: (width: number) => width > 2 },
+  xyzw: { wgsl: 'let r : vec4<T> = v.xyzw;', ok: (width: number) => width > 3 },
+  yxwz: { wgsl: 'let r : vec4<T> = v.yxwz;', ok: (width: number) => width > 3 },
+  rg: { wgsl: 'let r : vec2<T> = v.rg;', ok: true },
+  gr: { wgsl: 'let r : vec2<T> = v.gr;', ok: true },
+  rgg: { wgsl: 'let r : vec3<T> = v.rgg;', ok: true },
+  rgb: { wgsl: 'let r : vec3<T> = v.rgb;', ok: (width: number) => width > 2 },
+  grb: { wgsl: 'let r : vec3<T> = v.grb;', ok: (width: number) => width > 2 },
+  rgbr: { wgsl: 'let r : vec4<T> = v.rgbr;', ok: (width: number) => width > 2 },
+  rgba: { wgsl: 'let r : vec4<T> = v.rgba;', ok: (width: number) => width > 3 },
+  gbra: { wgsl: 'let r : vec4<T> = v.gbra;', ok: (width: number) => width > 3 },
+
+  // swizzle chains
+  xy_yx: { wgsl: 'let r : vec2<T> = v.xy.yx;', ok: true },
+  xyx_xxy: { wgsl: 'let r : vec3<T> = v.xyx.xxy;', ok: true },
+  xyz_zyx: { wgsl: 'let r : vec3<T> = v.xyz.zyx;', ok: (width: number) => width > 2 },
+  xyxy_rrgg: { wgsl: 'let r : vec4<T> = v.xyxy.rrgg;', ok: true },
+  rbrg_xyzw: { wgsl: 'let r : vec4<T> = v.rbrg.xyzw;', ok: (width: number) => width > 2 },
+  xyxz_rbg_yx: { wgsl: 'let r : vec2<T> = v.xyxz.rbg.yx;', ok: (width: number) => width > 2 },
+  wxyz_bga_xy: { wgsl: 'let r : vec2<T> = v.wxyz.bga.xy;', ok: (width: number) => width > 3 },
+
+  // error: invalid convenience letterings
+  xq: { wgsl: 'let r : vec2<T> = v.xq;', ok: false },
+  py: { wgsl: 'let r : vec2<T> = v.py;', ok: false },
+
+  // error: mixed convenience letterings
+  xg: { wgsl: 'let r : vec2<T> = v.xg;', ok: false },
+  ryb: { wgsl: 'let r : vec3<T> = v.ryb;', ok: false },
+  xgza: { wgsl: 'let r : vec4<T> = v.xgza;', ok: false },
+
+  // error: invalid index value
+  literal_5: { wgsl: 'let r : T = v[5];', ok: false },
+  literal_minus_1: { wgsl: 'let r : T = v[-1];', ok: false },
+
+  // error: invalid index type
+  float_idx: { wgsl: 'let r : T = v[1.0];', ok: false },
+  bool_idx: { wgsl: 'let r : T = v[true];', ok: false },
+  array_idx: { wgsl: 'let r : T = v[array<i32, 2>()];', ok: false },
+};
+
+g.test('vector')
+  .desc('Tests validation of vector indexed and swizzles')
+  .params(u =>
+    u
+      .combine('case', keysOf(kCases)) //
+      .combine('vector_width', [2, 3, 4] as const)
+      .combine('element_type', ['i32', 'u32', 'f32', 'f16', 'bool'] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.element_type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const c = kCases[t.params.case];
+    const enables = t.params.element_type === 'f16' ? 'enable f16;' : '';
+    const code = `${enables}
+
+alias T = ${t.params.element_type};
+
+struct S {
+  i : i32,
+}
+
+@compute @workgroup_size(1)
+fn main() {
+  var v : vec${t.params.vector_width}<T>;
```

Comment:
Worth parameterizing the source vector for `var` vs `let/param` vs `const`?

Reply:
Done